### PR TITLE
Added atomic checkpoint creation

### LINKED ISSUE #688: Checkpoint saving isn't atomic
## 🐛 Bug

Saving checkpoints happens non-atomically. In some cases, this causes an incomplete write of a checkpoint (for example when receiving a SIGKILL during writing), causing any subsequent loading to fail with 
`RuntimeError: unexpected EOF, expected 8 more bytes. The file might be corrupted.`

### To Reproduce

This is difficult to reproduce, since it relies on timing outside of code. For me, it happens with fast-running models that run at ~1-4 seconds per epoch.

### Expected behavior

Checkpointing should be resistant to such issues, and instead simply continue as-is.

### PR DESCRIPTION
**Pull request by fdiehl**
# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
  - #688
- [x] Did you read the [contributor guideline](https://github.com/williamFalcon/pytorch-lightning/blob/master/.github/CONTRIBUTING.md)?
- [x] Did you make sure to update the docs?   
  - No visible changes
- [x] Did you write any new necessary tests?  
  - Old behaviour continues; new behaviour relies on timing outside the code.

## What does this PR do?
Fixes #688 .

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

### CODE REVIEW

**Comment by williamFalcon** [comment]
@fdiehl can  you rebase master?

**Comment by fdiehl** [comment]
@williamFalcon Done.

(I had originally used an older commit because the tests previously failed on master due to tensorboard issues. They are passing with the current master.)

File: pytorch_lightning/trainer/training_io.py [hunk 1]
```diff
@@ -255,17 +255,22 @@ def term_handler(self, signum, frame):
     # --------------------
     # MODEL SAVE CHECKPOINT
     # --------------------
+    def _atomic_save(self, checkpoint, filepath):
```
[the remark this answers is not in the document]
Comment (fdiehl) [reply pytorch_lightning/trainer/training_io.py]: It's a file, as a replacement of `torch.save` . I've documented it.

[the remark this answers is not in the document]
Comment (fdiehl) [reply pytorch_lightning/trainer/training_io.py]: Excellent point. I've added docs.

**Pull request merged**